torch/_decomp/__init__.py

```diff
@@ -250,13 +252,172 @@ def remove_decompositions(
+
+    if is_mutating_or_aliasing:
+        return False
+
+    if not torch._C._dispatch_has_kernel(op_overload.name()):
+        return False
+
+    return True
+
+
+def _is_cia_op(op: "OpOverload") -> bool:
+    return (
+        torch._C._dispatch_has_kernel_for_dispatch_key(
+            op.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        )
+        or torch._C.DispatchKey.CompositeImplicitAutograd in op.py_kernels
+    )
+
+
+def _is_special_op_to_decompose(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    # Can't call kernel.decompose due to infinite recursion as
+    # we register this kernel to py_impl directly
+    dk = torch._C.DispatchKey.CompositeImplicitAutograd
+    if torch._C._dispatch_has_kernel_for_dispatch_key(
+        kernel.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+    ):
+        return kernel._op_dk(dk, *args, **kwargs)
+    else:
+        return NotImplemented
+
+
+def _is_special_op_to_preserve(*args, **kwargs):
+    return NotImplemented
+
+
+def _collect_all_valid_cia_ops():
+    cia_ops = torch._C._dispatch_get_registrations_for_dispatch_key(
+        "CompositeImplicitAutograd"
+    )
+    # Ignore quantized namespace ops
+    cia_ops = [name[6:] for name in cia_ops if name.startswith("aten::")]
+    # Materialize all CIA ops first
+    for op in cia_ops:
+        split_list = op.split(".")
+        # Sometime overload could be missing
+        assert len(split_list) == 1 or len(split_list) == 2
+        op_name = split_list[0]
+        op_overload_name = "default"
+        if len(split_list) == 2:
+            op_overload_name = split_list[1]
+
+        _ = getattr(getattr(torch.ops.aten, op_name), op_overload_name)
+
+    cia_ops = set()
+    for op in torch.ops.aten:
+        op_packet = getattr(torch.ops.aten, op)
+        for overload in op_packet.overloads():
+            op_overload = getattr(op_packet, overload)
+            if _check_valid_to_preserve(op_overload) and _is_cia_op(op_overload):
+                cia_ops.add(op_overload)
+    return cia_ops
+
+
 # See NOTE [Core ATen Ops]
 #
 # list was copied from torch/_inductor/decomposition.py
 # excluding decompositions that results in prim ops
 # Resulting opset of decomposition is core aten ops
 def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
+    decomp_table = _core_aten_decompositions_post_autograd()
+
+    # If it is fbcode change, we return the old decomposition list
+    from torch._inductor import config
+
+    if config.is_fbcode():
+        return decomp_table
+
+    aten = torch.ops.aten
+
+    # We are deleting custom decomp in core_aten_decomp
+    # for CIA ops but it should be fine technically
+    # because this table is only meant to be used in export context
+    # in which we really carefully control the decomp behaviour
+    # In any case, C++ decomps should be preferred
+    cia_ops_that_should_be_removed = [
```

Comment:
oh my comment was more: instead of doing this, why not just delete these entries from the source code? e.g. here: https://github.com/pytorch/pytorch/blob/main/torch/_decomp/__init__.py#L269

we have `aten.all` in the core aten decomp list.  `aten.dimname` is CIA (wrong to put into the table), but the other overloads seem fine (`aten.all.dim`, `aten.all.dims`). So we could alternatively modify the core aten decomp list directly to only have those valid overloads.

If you want to do that now vs. in a followup, either sounds ok

Reply:
I will do it in follow up because we do need to seperate out the fbcode only changes first.

Reply:
Tracking issue: https://github.com/pytorch/pytorch/issues/135759